src/main.rs

```diff
@@ -1,21 +1,46 @@
 use anyhow::{Context, Result};
 use rusb;
-use rusb::{DeviceHandle, Language, UsbContext};
+use rusb::{DeviceHandle, Direction, Language, Recipient, RequestType, UsbContext};
 use std::path::{Path, PathBuf};
 
-fn utf16_to_be_string(s: &str) -> String {
-    s.encode_utf16()
-        .map(|b| format!("{:04x}", b.to_be()))
-        .collect()
+fn bytestring(s: &[u8]) -> String {
+    s.iter()
+        .map(|b| format!("{b:02x}"))
+        .collect::<Vec<String>>()
+        .join("")
 }
 
-fn string_descriptor(
+/// The hid-uclogic driver queries the bytes for this buffer
+/// and returns the *full* buffer, including the len + DT prefix bytes.
+/// To be compatible with that code, let's do the same here.
+fn huion_string_descriptor(
     handle: &DeviceHandle<rusb::Context>,
     lang: &Language,
     index: u8,
-) -> rusb::Result<String> {
+) -> rusb::Result<Vec<u8>> {
     let timeout = std::time::Duration::from_millis(100);
-    handle.read_string_descriptor(*lang, index, timeout)
+
+    // This matches rusb::DeviceHandle::read_string_descriptor() but
+    // that function enforces a utf16 string with even lengths.
+    // But we don't have a string here, we just have random bytes,
+    // and on the Kamvas 12 our return buffer is length 19 which
+    // always results in Error::BadDescriptor.
+    let mut buf = [0u8; 256];
+    let len = handle.read_control(
+        rusb::request_type(Direction::In, RequestType::Standard, Recipient::Device),
+        rusb::constants::LIBUSB_REQUEST_GET_DESCRIPTOR,
+        u16::from(rusb::constants::LIBUSB_DT_STRING) << 8 | u16::from(index),
+        lang.lang_id(),
+        &mut buf,
+        timeout,
+    )?;
+    // buf[0] is length of the buffer
+    // buf[1] is the descriptor type (LIBUSB_DT_STRING == 0x3)
+    if buf[0] != len as u8 {
+        return Err(rusb::Error::BadDescriptor);
+    }
+
+    Ok(buf[..len].to_vec())
```

Comment:
I won't argue too much here, but I'm not sure there is a point in returning the first 2 bytes. They are USB implementation detail, and the payload is actually `2..len`.

But OTOH, this allows to check from udev that the string is complete, so meh?

Reply:
there's no point since they're effectively fixed but this is solely to match the data buffer with the one the kernel driver uses so we have the same offsets/indices. which I think might be useful long-term.